Modularization of loader

dylandreimerink opened the pull request:
This PR is the first in an effort to "clean up" the loader. This work un-exports all internal functions and types and exposes only the externally used functions via the Loader interface. This interface is implemented by a loader struct and is available as cell. This will allow us to make the loader dependent on other functionality and be depended upon, thus going from implicit dependencies to explicit dependencies.

```release-note
Modularization of loader
```

dylandreimerink committed:
loader: Unexport functions and types where possible

Most of the functions and types in the loader package are only used
internally. This commit unexports them. Some functions were placed in
the loader package but without much added benefit.

These changes make it easier to identify the API boundary of the loader
package.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
loader: Make remaining global functions methods of the loader object

Not that all global functions are methods of the loader object, we can
provide the loader via hive.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
loader: Make loader into cell and provide to hive

This commit changes the loader into a cell. The Datapath object depends
on it and re-distributes it still (for now). All locations where a
new loader was created are now replaced by the datapath interface so
there is now only ever one loader object at a time.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
datapath/loader: Moved compilation lock into loader

Moved the compilation lock into the loader as part of the effort to
separate the tight integration of the loader with the rest of the
agent and daemon.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

jschwinger233 reviewed (approved):
Makes sense overall.

lmb commented on daemon/cmd/daemon.go:
```diff
@@ -1075,8 +1075,16 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
-	if err := loader.RestoreTemplates(option.Config.StateDir); err != nil {
-		log.WithError(err).Error("Unable to restore previous BPF templates")
+	// Simplest implementation: Just garbage-collect everything.
+	// In future we should make this smarter.
```
I disagree: let's keep it dumb :laughing:

dylandreimerink replied:
Not my comment, I fully agree! Lets not make things smarter than they have to be.

lmb commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -448,8 +447,25 @@ func finishKubeProxyReplacementInit() error {
-			return fmt.Errorf("Cannot set NodePort acceleration: %w", err)
+		switch option.Config.NodePortAcceleration {
+		case option.XDPModeNative:
+			if option.Config.XDPMode == option.XDPModeLinkNone ||
```
This might read easier as a switch as well?

lmb commented on pkg/datapath/loader/hash.go:
```diff
@@ -14,10 +14,10 @@ import (
 	// code. See the definition of CILIUM_DATAPATH_SHA256 in Makefile.defs for
 	// details.
-	DatapathSHA256 string
+	datapathSHA256 string
```
Doesn't the makefile need updating if we change the name?

dylandreimerink replied:
Good catch!

lmb commented on pkg/datapath/types/loader.go:
```diff
@@ -25,6 +27,9 @@ type Loader interface {
 	Reinitialize(ctx context.Context, o BaseProgramOwner, tunnelConfig tunnel.Config, deviceMTU int, iptMgr IptablesManager, p Proxy) error
 	HostDatapathInitialized() <-chan struct{}
 	DeviceHasTCProgramLoaded(hostInterface string, checkEgress bool) (bool, error)
+	ELFSubstitutions(ep Endpoint) (map[string]uint64, map[string]string)
```
Could this type just be an alias for `loader.Loader`?

dylandreimerink replied:
No we can't, this type is duplicated to avoid an import cycle. One day, we will be able to remove it, but not yet.

lmb commented on pkg/datapath/loader/loader.go:
```diff
@@ -70,7 +70,12 @@ type loader struct {
 
-	ipsecMu lock.Mutex // guards reinitializeIPSec
+	ipsecMu         lock.Mutex    // guards reinitializeIPSec
+	compilationLock *lock.RWMutex // guards compiler usage
```
Why not without the pointer type for consistency with ipsecMu?

lmb reviewed (changes requested):
Why does `loader: Make remaining global functions methods of the loader object` make sense? Seems like the functions don't actually end up accessing state from the loader object. Maybe some more explanation in the commit message would help.

aditighag reviewed (approved):
Looks good for my codeowners.

lmb commented on Makefile.defs:
```diff
@@ -138,7 +138,7 @@ endif
 BPF_SRCFILES := $(filter-out $(BPF_SRCFILES_IGNORE),$(subst ../,,$(subst $(ROOT_DIR)/,,$(BPF_SRCFILES))))
 CILIUM_DATAPATH_SHA256=$(shell cd $(ROOT_DIR); cat $(BPF_SRCFILES) | sha256sum | awk '{print $$1}')
-GO_BUILD_LDFLAGS += -X "github.com/cilium/cilium/pkg/datapath/loader.DatapathSHA256=$(CILIUM_DATAPATH_SHA256)"
+GO_BUILD_LDFLAGS += -X "github.com/cilium/cilium/pkg/datapath/loader.datapathSHA256=$(CILIUM_DATAPATH_SHA256)"
```
Is it possible to add a unit test which asserts that this isn't empty or similar?

giorio94 committed:
gha: add step to ensure presence/absence of the AWS iptables chains

Recently, a regression of the poststart-eni.sh script got unnoticed,
as we are not explicitly testing it as part of the E2E pipelines.

Let's add a step to the Conformance EKS and Conformance AWS-CNI
workflows to assert that, in the former case, the stale AWS iptables
chains are removed, and in the latter they are not modified.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
gha: enable IPv6 in clustermesh upgrade/downgrade workflow

Now that known issues causing connection disruption (which appeared to
mostly affect dual stack clusters) have been fixed, let's enable IPv6
again in the clustermesh upgrade/downgrade workflow.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

mhofstetter committed:
ci: add documentation check to documentation workflow

Currently, differences in generated documentation are only checked
in the Travis workflow; and only due to a transitive dependency when
executing `make` in `.travis/build.sh`.

Therefore, this commit executes the checks in the documentation workflow.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
build: mark codeowners.rst as phony target

This commit marks the make target `codedowners.rst` (Documentation)
as phony target. This way, the generated gets triggered independent of
changes in the CODEOWNERS file.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
ci: don't execute postcheck on integration tests (travis & gha)

Currently, part of the integration test execution is executing
`make` in `.travis/build.sh`. The default target (`all`) implicitly
executes `precheck`, `build` & `postcheck`.

With the execution of the postcheck in the documentation workflow,
it's not longer necessary to execute them in the integrations too.

Therefore, this commit removes its execution.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

lmb committed:
complexity-tests: fix copying verifier log output to artifacts

It turns out that the following find command is subtle:

    find . -name foo -o -name bar -exec ...

This means: execute the command for bar. It doesn't execute the command for foo.
For this reason verifier logs are currently not being copied correctly into
GH artifacts.

Fixes: 735807fb16 ("test/verifier: fix complexity tests not being recompiled")
Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

jrajahalme committed:
envoy: Remove log filter that is no longer needed

While https://github.com/envoyproxy/envoy/issues/13504 is still open to
cover some cases, the reason for Envoy logs filtering seems to have been
fixed, so we can remove the logs filter for this.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Silence expected warning about internal listener feature

Silence the expected warning about internal listener being
work-in-progress. This special case can be removed when at Envoy 1.28,
as internal listener is a stable feature in Envoy 1.28.

CI Ginkgo tests use Envoy trace mode (debug-verbose=envoy) so we still
need to whitelist this warning.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter committed:
ingress: add unit tests to test default ingressclass

This commit adds a unit-test that tests the use case of the
Cilium IngressClass being the default or not.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

qmonnet committed:
ci/ipsec: Skip waiting for images when skipping upgrade/dowgrade

The workflow to test upgrades/downgrades with IPsec supports two cases:
upgrading from the last minor version, or from the last patch release.
In the case of patch releases, there are instances where we are not able
to determine the latest patch releases. This is the case on the current
development branch, where no patch releases have been created yet, for
example. In such a case, we skip most of the steps in the job.

When implementing the check on the patch release value and the bypass of
the relevant steps, we omitted to skip the step where we wait for the CI
images to be available, potentially causing unnecessary wait time for
the job. Let's skip this step as well. Since we're at it, we can also
skip the step for setting Kind parameters.

Fixes: c9dedb49f5a6 ("ci/ipsec: Skip upgrade/downgrade test to patch release on main branch")
Fixes: cd93d3774959 ("ci-ipsec-upgrade: Use lvh-kind")
Reported-by: Paul Chaignon <paul.chaignon@gmail.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

giorio94 committed:
renovate: enable Cilium CLI patch updates for Cilium <v1.14

Cilium versions earlier that v1.14 are officially compatible with the
Cilium CLI v0.14 series only. Still, some workflows in stable branches
are already using one of the v0.15 releases, to benefit from newer
functionalities. Let's enable renovate to bump patch versions for
them, so that we automatically pick new fixes and improvements.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

renovate[bot] committed:
chore(deps): update all github action dependencies to v5

Signed-off-by: renovate[bot] <bot@renovateapp.com>
